embulk-guess-msgpack/src/main/java/org/embulk/guess/msgpack/MsgpackGuessPlugin.java

```diff
@@ -0,0 +1,201 @@
+import org.embulk.spi.json.JsonArray;
+import org.embulk.spi.json.JsonObject;
+import org.embulk.spi.json.JsonValue;
+import org.embulk.util.config.ConfigMapperFactory;
+import org.embulk.util.guess.SchemaGuess;
+import org.embulk.util.msgpack.core.MessagePack;
+import org.embulk.util.msgpack.core.MessagePackException;
+import org.embulk.util.msgpack.core.MessageTypeException;
+import org.embulk.util.msgpack.core.MessageUnpacker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MsgpackGuessPlugin implements GuessPlugin {
+    @Override
+    public ConfigDiff guess(final ConfigSource config, final Buffer sample) {
+        final ConfigDiff configDiff = CONFIG_MAPPER_FACTORY.newConfigDiff();
+
+        if (!"msgpack".equals(config.getNestedOrGetEmpty("parser").get(String.class, "type", "msgpack"))) {
+            return configDiff;
+        }
+
+        final ConfigSource parserConfig = config.getNestedOrGetEmpty("parser");
+        String fileEncodingVariable = parserConfig.get(String.class, "file_encoding", null);
+        String rowEncodingVariable = parserConfig.get(String.class, "row_encoding", null);
+
+        try {
+            if (fileEncodingVariable == null || rowEncodingVariable == null) {
+                final MessageUnpacker uk = newUnpacker(sample);
+                try {
+                    uk.unpackArrayHeader();
+                    try {
+                        uk.unpackArrayHeader();
+                        fileEncodingVariable = "array";
+                        rowEncodingVariable = "array";
+                    } catch (final MessageTypeException ex) {
+                        fileEncodingVariable = "sequence";
+                        rowEncodingVariable = "array";
+                    }
+                } catch (final MessageTypeException ex) {
+                    final MessageUnpacker uk2 = newUnpacker(sample);  // TODO unpackArrayHeader consumes buffer (unexpectedly)
+                    try {
+                        final int n = uk.unpackMapHeader();
+                        fileEncodingVariable = "sequence";
+                        rowEncodingVariable = "map";
+                    } catch (final MessageTypeException ex2) {
+                        return configDiff;
+                    }
+                }
+            }
+
+            final String fileEncoding = fileEncodingVariable;
+            final String rowEncoding = rowEncodingVariable;
+
+            final MessageUnpacker uk = newUnpacker(sample);
+
+            if (fileEncoding.equals("array")) {
+                uk.unpackArrayHeader();  // skip array header to convert to sequence
+            } else if (fileEncoding.equals("sequence")) {
+                // do nothing
+            }
+
+            final List<ConfigDiff> schema;
+
+            if (rowEncoding.equals("map")) {
+                final ArrayList<LinkedHashMap<String, Object>> rows = new ArrayList<>();
+                try {
+                    while (true) {
```

Comment:
It seems that `IllegalStateException` isn't a subclass of the `IOException`. 

Is this part completely the same as the original?

```ruby
        begin
          while true
            rows << JSON.parse(uk.unpackValue.toJson)
          end
        rescue java.io.EOFException
        end
```

Reply:
Oh, I understand this part after I commented, that the parser raises `IOException` when parsing reaches the end of the part. So, This is completely the same original part.

Reply:
Yeah, this `IOException` seems like an "expected" Exception -- it's a bad pattern, but unavoidable for such a guessing logic.